[WPE] Add support for generating breakpad dumps

lauromoura opened the pull request:
#### b4d401778976615f04c35a1285c8749d75b5a974
<pre>
[WPE] Add support for generating breakpad dumps
<a href="https://bugs.webkit.org/show_bug.cgi?id=239103">https://bugs.webkit.org/show_bug.cgi?id=239103</a>

Reviewed by Adrian Perez de Castro.

Rebase of original patch from Vivek Arumugam from
<a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/808">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/808</a>

Breakpad is a library and tool suite that allows collection of
crash dumps in a compact format from applications stripped of
of debug information at runtime. It works by processing the debug
information generated in the build before it&apos;s stripped and
processing together with the minidump.

To enable, use the cmake option `-DENABLE_BREAKPAD`. The directory
where the minidumps will be stored can be set with `-DBREAKPAD_MINIDUMP_DIR`
cmake build option and the `$BREAKPAD_MINIDUMP_DIR` environment
variable.

In a followup commit, SIGPIPE should be ignored when using Breakpad as
it may interfere with generating the minidumps.

* Source/WebKit/PlatformWPE.cmake: Import the Breakpad library
* Source/WebKit/Shared/AuxiliaryProcessMain.h: Declare
AuxProcMainCommon. It&apos;s used to run initialization code common to all
subprocesses before WebKit itself is initialized.
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
(WebKit::AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon):
Install Breakpad handler if enabled beforey anything else.
* Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp: Added.
(WebKit::installBreakpadExceptionHandler):
* Source/WebKit/Shared/unix/BreakpadExceptionHandler.h: Added.
* Source/WebKit/Shared/win/AuxiliaryProcessMainWin.cpp:
(WebKit::AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon):
Default impl.
* Source/cmake/FindBreakpad.cmake: Added.
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake: Expose Breakpad option to other
ports.
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.setup_environ_for_server): Forward BREAKPAD_MINIDUMP_DIR if
set.

Canonical link: <a href="https://commits.webkit.org/253381@main">https://commits.webkit.org/253381@main</a>
</pre>

lauromoura commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,62 @@
+        return;
+
+    if (FileSystem::fileType(breakpadMinidumpDir) != FileSystem::FileType::Directory) {
+        fprintf(stderr, "Breakpad dir \"%s\" is not a directory, not installing handler\n", breakpadMinidumpDir.utf8().data());
```
Replaced the `WTFLogAlways` from previous iteration as when this code is run WebKit is not yet initialized.

aperezdc replied:
You should keep the `WTFLogAlways()`, because using it does _not_ need any WebKit initialization: it ends up calling `vfprintf(stderr, ...)` in our ports (see `Source/WTF/wtf/Assertions.cpp`, line 183).

lauromoura replied:
Oops. Didn't know that. Thanks.

aperezdc commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,62 @@
+        return;
+
+    if (FileSystem::fileType(breakpadMinidumpDir) != FileSystem::FileType::Directory) {
+        fprintf(stderr, "Breakpad dir \"%s\" is not a directory, not installing handler\n", breakpadMinidumpDir.utf8().data());
```
We can use `WTFLogAlways()`, as it's guaranteed to always work even if WebKit's initialization has not been called.

aperezdc reviewed (approved):
LGTM, but please do change back to `WTFLogAlways()` before landing.

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,62 @@
+/*
+ * If not stated otherwise in this file or this component's license file the
+ * following copyright and licenses apply:
```
It's almost as if whoever wrote this text designed it to be maximally confusing. I think you're supposed to remove this when copying code from wherever this used to live to someplace new. Otherwise, the implication is that you can ignore the Apache license and apply WebKit's BSD plus LGPLv2+ license instead, which probably isn't right?

Note that Apache is not actually a permitted license for WebKit, but we do already use it in a few places, so adding a little more doesn't make much difference. I'm unsure what to do about that.

mcatanzaro commented on Source/cmake/OptionsWPE.cmake:
```diff
@@ -119,6 +119,7 @@ if (ENABLE_DEVELOPER_MODE)
     WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_API_TESTS PRIVATE ON)
     WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_MINIBROWSER PUBLIC ON)
     WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_COG PRIVATE ON)
+    WEBKIT_OPTION_DEFAULT_PORT_VALUE(ENABLE_BREAKPAD PUBLIC ON)
```
Adrian, can you confirm you want this ON by default?

aperezdc replied:
Ahhh, good catch, I would actually prefer that this is `OFF` by default, to not add a new dependency which is not always packaged in distros (and whoever needs Breakpad support in an embeded device _should_ know what they are up to and how to enable it.)

aperezdc reviewed (changes requested):
@lauromoura Let's make the option `OFF` by default, please 🙏🏼

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,59 @@
+ */
+
+#if ENABLE(BREAKPAD)
+#include "config.h"
```
You have to `#include "config.h"` first for `ENABLE()` and `ENABLE_BREAKPAD`. This probably only works due to unified build?

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,59 @@
+#include "config.h"
+#include "BreakpadExceptionHandler.h"
+
+#include <client/linux/handler/exception_handler.h>
```
Weird include path. It seems like a clear error in breakpad.pc.in and breakpad-client.pc.in, [here](https://github.com/google/breakpad/blob/defdcb714e0c8113ade899e416d3024ac8024548/breakpad-client.pc.in#L4). `includedir=@includedir@/@PACKAGE_NAME@` should be `includedir=@includedir@`, that way you `#include <breakpad/client/linux/handler/exception_handler.h>` instead of the very confusing `<client/linux/handler/exception_handler.h>`. I wonder if Google would be interested in fixing that, at the cost of breaking current API users.

Anyway, I think you can get away with including it with writing `#include <breakpad/client/linux/handler/exception_handler.h>` anyway, since the top-level include directory will inevitably be in the include path regardless.

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,59 @@
+#include <wtf/FileSystem.h>
+#include <wtf/NeverDestroyed.h>
+
+
```
```suggestion

```

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,59 @@
+
+namespace WebKit {
+
+static std::once_flag onceFlag;
```
Put this inside the function.

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void installBreakpadExceptionHandler()
+{
+    static WTF::MainThreadLazyNeverDestroyed<google_breakpad::ExceptionHandler> exceptionHandler;
```
```suggestion
    static MainThreadLazyNeverDestroyed<google_breakpad::ExceptionHandler> exceptionHandler;
```

mcatanzaro commented on Source/WebKit/Shared/unix/BreakpadExceptionHandler.h:
```diff
@@ -0,0 +1,28 @@
+
+namespace WebKit {
+
+WK_EXPORT void installBreakpadExceptionHandler();
```
Uh-oh, this is a problem. This is an internal API: it is not exposed in the WPE public API. The `WK_EXPORT` here makes it accessible to tools and tests when running from the build directory, not to applications; when installed, the symbol will stripped away by our linker version script (webkitglib-symbols.map).

This needs a rethink. Does the application really need to call this API itself? Probably better for WebKit to handle this itself?

mcatanzaro replied:
Well, sorry, I see you do use it in AuxiliaryProcessMain. So just remove the `WK_EXPORT` and we're good:

```suggestion
void installBreakpadExceptionHandler();
```

lauromoura committed:
[WPE] Add support for generating breakpad dumps
https://bugs.webkit.org/show_bug.cgi?id=239103

Reviewed by Adrian Perez de Castro.

Rebase of original patch from Vivek Arumugam from
https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/808

Breakpad is a library and tool suite that allows collection of
crash dumps in a compact format from applications stripped of
of debug information at runtime. It works by processing the debug
information generated in the build before it's stripped and
processing together with the minidump.

To enable, use the cmake option `-DENABLE_BREAKPAD`. The directory
where the minidumps will be stored can be set with `-DBREAKPAD_MINIDUMP_DIR`
cmake build option and the `$BREAKPAD_MINIDUMP_DIR` environment
variable.

In a followup commit, SIGPIPE should be ignored when using Breakpad as
it may interfere with generating the minidumps.

* Source/WebKit/PlatformWPE.cmake: Import the Breakpad library
* Source/WebKit/Shared/AuxiliaryProcessMain.h: Declare
AuxProcMainCommon. It's used to run initialization code common to all
subprocesses before WebKit itself is initialized.
* Source/WebKit/Shared/unix/AuxiliaryProcessMain.cpp:
(WebKit::AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon):
Install Breakpad handler if enabled beforey anything else.
* Source/WebKit/Shared/unix/BreakpadExceptionHandler.cpp: Added.
(WebKit::installBreakpadExceptionHandler):
* Source/WebKit/Shared/unix/BreakpadExceptionHandler.h: Added.
* Source/WebKit/Shared/win/AuxiliaryProcessMainWin.cpp:
(WebKit::AuxiliaryProcessMainCommon::AuxiliaryProcessMainCommon):
Default impl.
* Source/cmake/FindBreakpad.cmake: Added.
* Source/cmake/OptionsWPE.cmake:
* Source/cmake/WebKitFeatures.cmake: Expose Breakpad option to other
ports.
* Tools/Scripts/webkitpy/port/wpe.py:
(WPEPort.setup_environ_for_server): Forward BREAKPAD_MINIDUMP_DIR if
set.

Canonical link: https://commits.webkit.org/253381@main
